pkg/workloads/spark_job/spark_util.py

```diff
@@ -220,53 +229,61 @@ def ingest(ctx, spark):
-            log_df_schema(df, logger.error)
-
-            raise UserException(
-                "raw column " + raw_column_name,
-                "type mismatch",
-                "expected {} but found {}".format(
-                    " or ".join(str(x) for x in expected_types), actual_type
-                ),
-            )
-        target_type = CORTEX_TYPE_TO_SPARK_TYPE[raw_column["type"]]
+        expected_cortex_type = raw_column["type"]
+        if expected_cortex_type == consts.COLUMN_TYPE_INFERRED:
+            column_type = SPARK_TYPE_TO_CORTEX_TYPE[inferred_col_type_map[raw_column_name]]
+            ctx.write_metadata(raw_column["id"], {"type": column_type})
+        else:
+            actual_spark_type = input_type_map[raw_column_name]
+            if actual_spark_type not in SPARK_TYPE_TO_CORTEX_TYPE:
+                actual_spark_type = StringType()
+
+            expected_types = CORTEX_TYPE_TO_CASTABLE_SPARK_TYPES[expected_cortex_type]
+            if actual_spark_type not in expected_types:
+                logger.error("found schema:")
+                log_df_schema(df, logger.error)
+
+                raise UserException(
+                    "raw column " + raw_column_name,
+                    "type mismatch",
+                    "expected {} but found {}".format(
+                        " or ".join(str(x) for x in expected_types), actual_spark_type
+                    ),
+                )
 
-        if target_type != actual_type:
-            df = df.withColumn(raw_column_name, F.col(raw_column_name).cast(target_type))
+            if actual_spark_type != expected_cortex_type:
+                target_spark_type = CORTEX_TYPE_TO_SPARK_TYPE[expected_cortex_type]
+                df = df.withColumn(raw_column_name, F.col(raw_column_name).cast(target_spark_type))
 
     return df.select(*sorted(df.columns))
 
 
 def read_csv(ctx, spark):
     data_config = ctx.environment["data"]
-    expected_field_names = data_config["schema"]
-
-    schema_fields = []
-    for field_name in expected_field_names:
-        if field_name in ctx.raw_columns:
-            spark_type = CORTEX_TYPE_TO_SPARK_TYPE[ctx.raw_columns[field_name]["type"]]
-        else:
-            spark_type = StringType()
-
-        schema_fields.append(StructField(name=field_name, dataType=spark_type))
 
     csv_config = {
         util.snake_to_camel(param_name): val
         for param_name, val in data_config.get("csv_config", {}).items()
         if val is not None
     }
 
-    df = spark.read.csv(
-        data_config["path"], schema=StructType(schema_fields), mode="FAILFAST", **csv_config
-    )
-    return df.select(*ctx.raw_columns.keys())
+    df = spark.read.csv(data_config["path"], inferSchema=True, mode="FAILFAST", **csv_config)
+    renamed_cols = [F.col(c).alias(data_config["schema"][idx]) for idx, c in enumerate(df.columns)]
+    df = df.select(*renamed_cols)
+
+    if set(data_config["schema"]) != set(df.columns):
```

Comment:
Does this check make sense, since the columns are being renamed on line 274? Maybe we should just check the number of ingested columns matches the number of columns in the schema, and complain if not equal?

Try hitting this error as a user would and let me know what you think.

Reply:
Spark would assign default names (i.e. `c1,c2,c3`) when `inferSchema` is True, so we rename it back to the column names the user provided. This check catches the case when we're missing columns. I'll be refactor it to be more clear.